File: src/main/webapp/app/entities/ext-task/ext-task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpResponse, HttpRequest, HttpEvent } from '@angular/common/http';
import { Observable } from 'rxjs/Observable';
import { SERVER_API_URL } from '../../app.constants';

import { JhiEventManager, JhiParseLinks, JhiAlertService } from 'ng-jhipster';
import { JhiDateUtils } from 'ng-jhipster';
import { createRequestOption } from '../../shared';

import {SnFile} from '../sn-file/sn-file.model';

import { TaskGroup } from '../task-group/task-group.model';
import { Task } from '../task/task.model';
import { TaskHistory } from '../task-history/task-history.model';
import {TaskGroupService} from '../task-group/task-group.service';
import {TaskService} from '../task/task.service';

@Injectable()
export class ExtTaskService {

  private resourceUrl =  SERVER_API_URL + 'api/ext/';

  constructor(private http: HttpClient,
              private dateUtils: JhiDateUtils,
              private taskGroupService: TaskGroupService,
              private taskService: TaskService,
              private jhiAlertService: JhiAlertService) { }

  // get all the taskGroup falling in this date range.
  queryTaskGroupsByDate(req?: any): Observable<HttpResponse<TaskGroup[]>> {
    const options = createRequestOption(req);
    return this.http.get<TaskGroup[]>(this.resourceUrl + 'task-groups', { params: options, observe: 'response' })
        .map((res: HttpResponse<TaskGroup[]>) => this.taskGroupService.convertArrayResponse(res));
  }

  /*
  // get all the taskGroup falling in this date range.
  queryTaskGroupsByDate(req?: any): Observable<any> {
    const options = createRequestOption(req);
    return this.http.get(this.resourceUrl + 'task-groups', { params: options, observe: 'response' });
  }
  */

   // get all the tasks of taskGroup or active task or all task list, falling in this date range.
  // the kind of task list depends on the source param
  queryTasks(req?: any): Observable<HttpResponse<Task[]>> {
    const options = createRequestOption(req);
    return this.http.get<Task[]>(this.resourceUrl + 'tasks', { params: options, observe: 'response' })
        .map((res: HttpResponse<Task[]>) => this.taskService.convertArrayResponse(res));
  }

  /*
  // get all the tasks of taskGroup or active task or all task list, falling in this date range.
  // the kind of task list depends on the source param
  queryTasks(req?: any): Observable<any> {
    // const reqObj = {'map': req};
    const options = createRequestOption(req);
    return this.http.get(this.resourceUrl + 'tasks', { params: options, observe: 'response' });
        // .map((res: HttpResponse<Task[]>) => this.taskService.convertArrayResponse(res));
  }
  */

 countTaskStatus(taskGroupId: number): Observable<any> {
   return this.http.get(this.resourceUrl + 'count-of-status/' + taskGroupId);
 }

  fetchBundle(): Observable<any> {
    return this.http.get(this.resourceUrl + 'bundle');
  }

  fetchLoggedInUserInfo(): Observable<any> {
    return this.http.get(this.resourceUrl + 'login-user-info');
  }

  /**
   * clone given tasks.  Equal number of tasks created
   */
  cloneTasks(tasks: Task[]): Observable<any> {
    return this.http.post(this.resourceUrl + 'clone', tasks);
  }

  /**
   * merge given tasks.  One new task is created with data from first given task and files from all given tasks
   * all given tasks are marked as merged and inactive
   */
  mergeTasks(tasks: Task[]): Observable<any> {
    return this.http.post(this.resourceUrl + 'merge', tasks);
  }

  // upload all the files
  // source can be task or taskGroup
  // id can be zero or any id
  // if taskGroup and id is zero, a new taskGroup will be created
  // if task and id is zero, error is thrown
  pushFilesToStorage(source: string, id: number, fileList: FileList): Observable<HttpEvent<{}>> {

    if (fileList.length <= 0) {
      return null;
    }

    const formdata: FormData = new FormData();

    // formdata.append('file', fileList);

    for (let i = 0; i < fileList.length; i++) {
      formdata.append('file[]', fileList.item(i));
    }

    const finalUrl: string = this.resourceUrl + 'upload-files/' + source + '/' + id;
    console.log(finalUrl);
    const req = new HttpRequest('POST', finalUrl, formdata, {
      reportProgress: true,
      responseType: 'text'
    });

    return this.http.request(req);
  }

  getDownloadLink(source: string, id: number, selectedTasks: Task[]): string {
    let selectedIds = '';

    if (selectedTasks != null) {
      selectedTasks.forEach((task) => {
        selectedIds += task.id + ',';
      });
    }

    const url = this.resourceUrl + 'download-files/' + source + '/' + id + '/' +
           (selectedIds == null || selectedIds === '' ? '0' : selectedIds);

    // console.log(url);
    return url;
  }

  downloadFiles(source: string, id: number, selectedTasks: Task[]): Observable<any> {
    const finalUrl: string = this.getDownloadLink(source, id, selectedTasks);
    return this.http.get(finalUrl, { responseType: 'blob' });
  }

  updateTemplateCount(tasks: Task[], templateCount: number): Observable<any> {
    return this.http.put(this.resourceUrl + 'update-template-count', tasks);
  }

  /**
   * update given list of snFiles
   * @param snFiles
   */
  updateSnFiles(snFiles: SnFile[]): Observable<any> {
    return this.http.put(this.resourceUrl + 'update-snfiles', snFiles);
  }

  /**
   * update given list of tasks
   * tasks - list of tasks to be updated
   * historyObe - history records of are created based on status and notes from historyObe
   * fieldNames - comma separated fields of task that are updated.  Only for display purpose, the fieldNames are not modified
   */
  updateTasks(tasks: Task[], historyObe: TaskHistory, fieldNames: String)
  : Observable<any> {

    const paramObj = {
      'tasks': tasks,
      'historyObe': historyObe,
      'fieldNames': fieldNames
    };

    return this.http.put(this.resourceUrl + 'update-tasks', paramObj);
  }

  findSnFiles(taskId: number): Observable<any> {
    return this.http.get(this.resourceUrl + 'snfiles-of-task/' + taskId);
  }

  findHistory(taskId: number): Observable<any> {
    return this.http.get(this.resourceUrl + 'history-of-task/' + taskId);
  }

  // convert any type to corresponding enum value string
  varToEnumStringVal(value: any, enumerator: any): string {
    const en: { [index: string]: any } = enumerator;

    if (!isNaN(parseInt(value, 10))) {
        return en[value];
    }

    const num = en[value];

    return en[num];
  }
}

File: src/main/java/com/neemshade/sniper/service/ExtTaskService.java
package com.neemshade.sniper.service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.neemshade.sniper.domain.SnFile;
import com.neemshade.sniper.domain.Task;
import com.neemshade.sniper.domain.TaskGroup;
import com.neemshade.sniper.domain.TaskHistory;
import com.neemshade.sniper.domain.User;
import com.neemshade.sniper.domain.UserInfo;
import com.neemshade.sniper.domain.enumeration.TaskStatus;
import com.neemshade.sniper.repository.CompanyRepository;
import com.neemshade.sniper.repository.DoctorRepository;
import com.neemshade.sniper.repository.HospitalRepository;
import com.neemshade.sniper.repository.PatientRepository;
import com.neemshade.sniper.repository.TaskGroupRepository;
import com.neemshade.sniper.repository.TaskHistoryRepository;
import com.neemshade.sniper.repository.UserInfoRepository;
import com.neemshade.sniper.repository.UserRepository;
import com.neemshade.sniper.security.SecurityUtils;

@Service
@Transactional
public class ExtTaskService {	
	
	private final Logger log = LoggerFactory.getLogger(ExtTaskService.class);
	
	@Autowired
    private TaskGroupRepository taskGroupRepository;
	
	@Autowired
	private TaskService taskService;
	
	@Autowired
	private SnFileService snFileService;
	
	@Autowired
	private TaskHistoryRepository taskHistoryRepository;
	
	@Autowired
	private UserRepository userRepository;
	
	@Autowired
	private UserInfoRepository userInfoRepository;
	
	@Autowired
	private CompanyRepository companyRepository;
	
	@Autowired
	private HospitalRepository hospitalRepository;
	
	@Autowired
	private DoctorRepository doctorRepository;
	
	@Autowired
	private PatientRepository patientRepository;
	
	public enum TASK_UPDATE_PARAM {
	    TASKS ("tasks"),
	    HISTORY_OBE ("historyObe"),
	    FIELD_NAMES ("fieldNames");
		
		private String field;
		
		public String getField()
		{
			return field;
		}
	    
	 // enum constructor - cannot be public or protected
	    private TASK_UPDATE_PARAM(String field)
	    {
	        this.field = field;
	    }
	}
	
	
	
	public  static enum BUNDLE_FIELD {
		USER ("USER"),
	    COMPANY ("COMPANY"),
	    HOSPITAL ("HOSPITAL"),
	    DOCTOR ("DOCTOR"),
	    PATIENT("PATIENT"),
	    OWNER ("OWNER"),
	    TRANSCRIPT ("TRANSCRIPT"),
	    EDITOR ("EDITOR"),
	    MANAGER ("MANAGER"),
	    DIRECTOR ("DIRECTOR");
		
		private String field;
		
		public String getField()
		{
			return field;
		}
	    
	 // enum constructor - cannot be public or protected
	    private BUNDLE_FIELD(String field)
	    {
	        this.field = field;
	    }
	}
	
	
	
	

    public ExtTaskService() {
    }
    
    /**
     * find the userInfo of the logged in user
     * @return
     * @throws Exception 
     */
    public UserInfo fetchLoggedInUserInfo() throws Exception {
    	try {
    		Optional<String> userLogin = SecurityUtils.getCurrentUserLogin();
    		log.info("ets login=" + userLogin.get());
    		Optional<User> user = userRepository.findOneByLogin(userLogin.get());
    		Optional<UserInfo> userInfo = userInfoRepository.findOneByUserId(user.get().getId());

    		return userInfo.get();
    	}
    	catch(Exception ex)
    	{
    		throw new Exception("Unable to find logged in user " + ex.getMessage());
    	}
	}

    // retrieve taskgroup in range given date range
	public Page<TaskGroup> findTaskGroupByDates(Instant fromDate, Instant toDate, Pageable pageable) {
		
		return taskGroupRepository.findAllByCreatedTimeBetweenOrderByCreatedTimeDesc(fromDate, toDate, pageable);
	}
	

	/**
	 * multi handling fn.  gets tasks based on source.  it could be taskgroup's tasks, active tasks or all tasks of user
	 * @param source - taskGroup or activeTasks or allTasks
	 * @param taskGroupId - applicable when source = taskGroup
	 * @param fromDate
	 * @param toDate
	 * @param pageable
	 * @return
	 * @throws Exception
	 */
	public Page<Task> findTasks(
			String source, Long taskGroupId, 
			Instant fromDate, Instant toDate, Pageable pageable) 
					throws Exception {
		if(source == null || "".equals(source))
			throw new Exception("Invalid source " + source);

		if(source.equals("taskGroup"))
			return taskService.findTasksOfTaskGroup(
				taskGroupId,
				fromDate,
				toDate,
				pageable);
		
		if(source.equals("activeTasks"))
			return taskService.findActiveTasksOfUser(
				fromDate,
				toDate,
				pageable);
		
		if(source.equals("allTasks"))
			return taskService.findAllTasksOfUser(
				fromDate,
				toDate,
				pageable);
		
		throw new Exception("Invalid source " + source);
	}
	
	
	public Map<BUNDLE_FIELD, Object> fetchBundle() {
		Map<BUNDLE_FIELD, Object> map = new HashMap<BUNDLE_FIELD, Object>();
		
		map.put(BUNDLE_FIELD.COMPANY, companyRepository.findAllByOrderByCompanyName());
		map.put(BUNDLE_FIELD.HOSPITAL, hospitalRepository.findAllByOrderByHospitalName());
		map.put(BUNDLE_FIELD.DOCTOR, doctorRepository.findAllByOrderByDoctorName());
//		map.put(BUNDLE_FIELD.PATIENT, patientRepository.findAllByOrderByPatientName());
		map.put(BUNDLE_FIELD.USER, userInfoRepository.findAllByOrderByEmpCode());
		
		return map;
	}

	/**
	 * duplicate all the tasks.  If 3 tasks are given, there will be 3 more copies of these given tasks
	 * @param tasks
	 * @return
	 * @throws Exception 
	 */
	public String cloneTasks(List<Task> tasks) throws Exception {
		
		if(tasks == null || tasks.size() <= 0)
        {
        	throw new Exception("Error! empty tasks");
        }
		
		TaskHistory historyObe = new TaskHistory();
		historyObe.setTaskStatus(TaskStatus.CREATED);
		
		for (Task givenTask : tasks) {
			Task newTask = createTask(givenTask.getTaskGroup(), givenTask.getPeckOrder() + 5, null);
			xerox(givenTask, newTask);
			xeroxFiles(givenTask, newTask);
			taskService.save(newTask);
			
			historyObe.setNotes("Cloned from Task " + givenTask.getId());
			createTaskHistory(newTask, historyObe);
		}

		return "{ \"msg\" : \"Cloned " + tasks.size() + " tasks \" }";
	}
	
	/**
	 * merge tasks into one task.  For eg, if 3 tasks are given, one new task is created
	 * the new task will have data from the first task out of the given list
	 * all 3 given tasks will be marked inActive with the status Merged
	 * @param tasks
	 * @return
	 * @throws Exception 
	 */
	public String mergeTasks(List<Task> tasks) throws Exception {
		if(tasks == null || tasks.size() <= 1)
        {
        	throw new Exception("Error! too few tasks");
        }
		
		Task firstTask = tasks.get(0);
		Task newTask = createTask(firstTask.getTaskGroup(), firstTask.getPeckOrder() + 5, null);
		xerox(firstTask, newTask);
		
		TaskHistory historyObe = new TaskHistory();
		historyObe.setTaskStatus(TaskStatus.MERGED);
		historyObe.setNotes("Merged into Task " + newTask.getId());
		
		String taskIds = "";
		
		for (Task givenTask : tasks) {
			
			xeroxFiles(givenTask, newTask);
			
			givenTask.setIsActive(false);
			givenTask.setTaskStatus(TaskStatus.MERGED);
			taskService.save(givenTask);
			createTaskHistory(givenTask, historyObe);
			
			taskIds += ("".equals(taskIds) ? "" : ", ") + givenTask.getId();
		}
		
		taskService.save(newTask);
		
		historyObe.setTaskStatus(TaskStatus.CREATED);
		historyObe.setNotes("Merged from Tasks " + taskIds);
		createTaskHistory(newTask, historyObe);

		return "{ \"msg\" : \"Merged " + tasks.size() + " tasks \" }";
	}

	

	/**
	 * copy data from source to dest
	 * @param sourceTask
	 * @param destTask
	 */
	private void xerox(Task sourceTask, Task destTask) {
		destTask.setTaskTitle(sourceTask.getTaskTitle());
		destTask.setTaskStatus(sourceTask.getTaskStatus());
		destTask.setHasPMSignedOff(sourceTask.isHasPMSignedOff());
		destTask.setIsActive(sourceTask.isIsActive());
		destTask.setNotes(sourceTask.getNotes());
		
		destTask.setCompany(sourceTask.getCompany());
		destTask.setOwner(sourceTask.getOwner());
		destTask.setTranscript(sourceTask.getTranscript());
		destTask.setEditor(sourceTask.getEditor());
		destTask.setManager(sourceTask.getManager());
		destTask.setDoctor(sourceTask.getDoctor());
		destTask.setHospital(sourceTask.getHospital());
	}
	
	/**
	 * copy files from sourceTask into destTask
	 * @param sourceTask
	 * @param destTask
	 */
	private void xeroxFiles(Task sourceTask, Task destTask) {
		List<SnFile> snFiles = snFileService.findSnFilesOfTask(sourceTask.getId());
		destTask.getSnFiles().addAll(snFiles);
	}
	

	/**
	 * create a new task.  The title is given by the filename
	 * @param taskGroup
	 * @param peckOrder
	 * @param snFile
	 * @return
	 * @throws Exception 
	 */
	public Task createTask(TaskGroup taskGroup, Integer peckOrder, SnFile snFile) throws Exception {
		Task task = new Task();
		
		task.setTaskGroup(taskGroup);
		task.setTaskTitle(snFile == null ? "" : snFile.getFileName() + "." + snFile.getFileExt());
		task.setTaskStatus(TaskStatus.CREATED);
		task.setCreatedTime(Instant.now());
		task.setManager(fetchLoggedInUserInfo());
		task.setHasPMSignedOff(false);
		task.setIsActive(false);
		task.setPeckOrder(peckOrder);
		task.setNotes("created task");
		Task newTask = taskService.save(task);
		
		createTaskHistory(newTask, TaskStatus.CREATED, newTask.getNotes());
		
		return newTask;
	}
	
	/**
	   * update given list of tasks,  Given map contains the following
	   * tasks - list of tasks to be updated
	   * historyObe - history records of are created based on status and notes from historyObe
	   * fieldNames - comma separated fields of task that are updated.  Only for display purpose, the fieldNames are not modified
	 * @throws Exception 
	   */
	public String updateTasks(List<Object> result) throws Exception {
//		if(paramObj == null)
//		{
//			throw new Exception("Invalid tasks list.  param is empty");
//		}
		
//		JsonFactory factory = new JsonFactory();
//	    ObjectMapper mapper = new ObjectMapper();
//	    TypeFactory typeFactory = mapper.getTypeFactory();
//	    MapType mapType = typeFactory.constructMapType(HashMap.class, String.class, Object.class);

//	    Map<String, Object> result = mapper.convertValue(paramObj, mapType);
	    
//	    org.springframework.boot.json.JsonParser springParser = JsonParserFactory.getJsonParser();
//	    Map<String, Object> result = springParser.parseMap(response.get);
		
//		List<Task> tasks = (List<Task>) result.get("tasks");
//		TaskHistory historyObe = null;
//		String fieldNames = null;
//		historyObe = (TaskHistory) result.get("historyObe");
//		fieldNames = (String) result.get("fieldNames");
		
		List<Task> tasks = (List<Task>) result.get(0);
		TaskHistory historyObe = (TaskHistory) result.get(1);
		String fieldNames = (String) result.get(2);
		
//		LinkedHashMap lhmTasks = (LinkedHashMap) result.get(0);
//		Set keys = lhmTasks.keySet();
//		for (Object object : keys) {
//			log.debug("in ets updateTasks " + object.toString());
//		}
//		List<Task> tasks = lhmTasks.get(keys.);
//		TaskHistory historyObe = (TaskHistory) result.get(1);
//		String fieldNames = (String) result.get(2);
//		
		if(historyObe == null)
		{
			historyObe = new TaskHistory();
			historyObe.setNotes("Unknown operation");
			historyObe.setTaskStatus(TaskStatus.SETTING);
		}
				
		for(Task task : tasks) {
			taskService.save(task);
			createTaskHistory(task, historyObe);
		}
//		
		return "Updated " + tasks.size() + " tasks with " + fieldNames + " properties";
	}
	
	/**
	 * if doctor or hospital info available, update ws and wos line count in the snFile
	 * @param task
	 * @param snFile
	 */
	public void adjustLineCount(Task task, SnFile snFile) {
		Integer templateCount = null;
		
		if(task != null && task.getDoctor() != null && task.getDoctor().getTemplateCount() != null) {
			templateCount = task.getDoctor().getTemplateCount();
		}
		else {
			if(task != null && task.getHospital() != null && task.getHospital().getTemplateCount() != null) {
				templateCount = task.getHospital().getTemplateCount();
			}
		}
		
		if(templateCount == null || templateCount == 0)
			return; // nothing to update
		
		snFile.setWsAdjustedLineCount(snFile.getWsActualLineCount() - templateCount);
		snFile.setWsFinalLineCount(snFile.getWsAdjustedLineCount());
		
		snFile.setWosAdjustedLineCount(snFile.getWosActualLineCount() - templateCount);
		snFile.setWosFinalLineCount(snFile.getWosAdjustedLineCount());
	}
	
	/**
	 * most likely the pm would have changed the final count.  update all the given snFiles
	 * @param snFiles
	 * @return
	 */
	public String updateSnFiles(List<SnFile> snFiles) {
		
		
		for(SnFile snFile : snFiles) {
			SnFile updatedSnFile = snFileService.save(snFile);
		}
//		
		return "{ \"msg\" : \"Updated " + snFiles.size() + " files \" }";
	}

	
	public String updateTasks(List<Task> tasks, TaskHistory historyObe, String fieldNames) throws Exception {
		if(historyObe == null)
		{
			historyObe = new TaskHistory();
			historyObe.setNotes("Unknown operation");
			historyObe.setTaskStatus(TaskStatus.SETTING);
		}
				
		for(Task task : tasks) {
			Task updatedTask = taskService.save(task);
			createTaskHistory(updatedTask, historyObe);
		}
//		
		return "{ \"msg\" : \"Updated " + tasks.size() + " tasks with " + fieldNames + " properties\" }";
	}
	
	/**
	 * key method to create history record
	 * @param task
	 * @param taskStatus
	 * @param notes
	 * @throws Exception 
	 */
	public void createTaskHistory(Task task, com.neemshade.sniper.domain.enumeration.TaskStatus taskStatus, String notes) throws Exception
	{
		TaskHistory taskHistory = new TaskHistory();
		taskHistory.setTask(task);
		taskHistory.setTaskStatus(taskStatus);
		taskHistory.setPunchTime(Instant.now());
		taskHistory.setNotes(notes);
		taskHistory.setUserInfo(fetchLoggedInUserInfo());
		
		taskHistoryRepository.save(taskHistory);
	}


	/**
	 * create a history for a given task and given historyObe - a template with status and notes
	 * @param task
	 * @param historyObe
	 * @throws Exception
	 */
	public void createTaskHistory(Task task, TaskHistory historyObe) throws Exception
	{
		createTaskHistory(task, historyObe.getTaskStatus(), historyObe.getNotes());
	}

	public List<TaskHistory> findHistoryOfTask(Long taskId) {
		return taskHistoryRepository.findByTaskIdOrderByPunchTimeDesc(taskId);
	}

	

		
}
